### ISSUE TITLE
C++ code classes generated disordered "A uses undefined class B"

### ISSUE THREAD

**Issue by Liftu** [ReClassNET/ReClass.NET#232]
Hello,

I'm trying to generate the C++ code from the structs I reversed but I get an error when I'm compiling the result.

ReClass.NET seems to generate the structs definitions in a random order. 
So if a class A refers to a class B but the latter is defined after the struct A in the generated C++ code, I get the error "A uses undefined class B" from Visual Studio compiler.

I have more than 2000 lines of generated C++ code so I can't rearrange them manually.

It's the first time that kind of stuff happen to me using ReClass.NET and I wonder if someone else already encountered the same problem or not.

**Comment by KN4CK3R** [comment]
The class definitions are not in random order. ReClass checks which class depends on which other class and orders them for the compiler to be happy. Did you build a loop (A -> B -> C -> A)?

Otherwise I need a minimal example to have a deeper look.

**Comment by Liftu** [comment]
Ok, I think I managed to locate the issue.
This only occurs on classes that are used in unions.

I made a ReClass project example that I'll attach to this comment as a [ZIP archive](https://github.com/ReClassNET/ReClass.NET/files/9283646/example.zip).

And here is the generated C++ code from ReClass.NET :
```c
// Created with ReClass.NET 1.2 by KN4CK3R

class E
{
public:
	int32_t intE; //0x0000
}; //Size: 0x0004

class A
{
public:
	int32_t intA; //0x0000
	float floatA; //0x0004
	union //0x0008
	{
		class B b; //0x0000
		class C c; //0x0000
	};
	class E e; //0x0010
}; //Size: 0x0014

class B
{
public:
	int32_t intB; //0x0000
}; //Size: 0x0004

class C
{
public:
	int32_t intC; //0x0000
	union //0x0004
	{
		class D d; //0x0000
	};
}; //Size: 0x0008

class D
{
public:
	int32_t intD; //0x0000
}; //Size: 0x0004
```

As you can see, only the class E has been defined on top because it is the only class used as is in another one (in class A).
But classes B, C and D are used in unions and thus are not declared on the correct position.

For information I use the latest Github release from 2019/04/15.

Thank you for your time.

**Comment by KN4CK3R** [comment]
Thank you for the example. Fixed in 45696a71

```cpp
// Created with ReClass.NET 1.2 by KN4CK3R

class B
{
public:
	int32_t intB; //0x0000
}; //Size: 0x0004
static_assert(sizeof(B) == 0x4);

class D
{
public:
	int32_t intD; //0x0000
}; //Size: 0x0004
static_assert(sizeof(D) == 0x4);

class C
{
public:
	int32_t intC; //0x0000
	union //0x0004
	{
		class D d; //0x0000
	};
}; //Size: 0x0008
static_assert(sizeof(C) == 0x8);

class E
{
public:
	int32_t intE; //0x0000
}; //Size: 0x0004
static_assert(sizeof(E) == 0x4);

class A
{
public:
	int32_t intA; //0x0000
	float floatA; //0x0004
	union //0x0008
	{
		class B b; //0x0000
		class C c; //0x0000
	};
	class E e; //0x0010
}; //Size: 0x0014
static_assert(sizeof(A) == 0x14);
```
